🧑‍🌾 rclpy test subscription test regression

### Generated by Generative AI

_No response_

### Operating System:

Ubuntu noble

### ROS version or commit hash:

Rolling

### RMW implementation (if applicable):

Connext

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

rclpy

### 'ros2 doctor --report' output

_No response_

### Steps to reproduce issue

1. Run a build in [Rolling nightly connext](https://build.ros2.org/job/Rci__nightly-connext_ubuntu_noble_amd64/)
2. See test regressions

### Expected behavior

Tests should pass

### Actual behavior

Tests are failing

### Additional information

Reference build:
- https://build.ros2.org/job/Rci__nightly-connext_ubuntu_noble_amd64/531/

Test regressions:
- [projectroot.test_subscription](https://build.ros2.org/job/Rci__nightly-connext_ubuntu_noble_amd64/531/testReport/junit/(root)/projectroot/test_subscription/)
- [rclpy.rclpy.test.test_subscription.test_on_new_message_callback](https://build.ros2.org/job/Rci__nightly-connext_ubuntu_noble_amd64/531/testReport/junit/rclpy.rclpy.test/test_subscription/test_on_new_message_callback/)

Log output:
```
self = <Mock id='133622991097488'>, args = (1,), kwargs = {}
msg = "Expected 'mock' to be called once. Called 0 times."

    def assert_called_once_with(self, /, *args, **kwargs):
        """assert that the mock was called exactly once and that that call was
        with the specified arguments."""
        if not self.call_count == 1:
            msg = ("Expected '%s' to be called once. Called %s times.%s"
                   % (self._mock_name or 'mock',
                      self.call_count,
                      self._calls_repr()))
>           raise AssertionError(msg)
E           AssertionError: Expected 'mock' to be called once. Called 0 times.

/usr/lib/python3.12/unittest/mock.py:955: AssertionError

During handling of the above exception, another exception occurred:

test_node = <rclpy.node.Node object at 0x798786eb0290>

    def test_on_new_message_callback(test_node) -> None:
        topic_name = '/topic'
        cb = Mock()
        sub = test_node.create_subscription(
            msg_type=Empty,
            topic=topic_name,
            qos_profile=10,
            callback=cb)
        pub = test_node.create_publisher(Empty, topic_name, 10)
        sub.handle.set_on_new_message_callback(cb)
        cb.assert_not_called()
        pub.publish(Empty())
>       cb.assert_called_once_with(1)
E       AssertionError: Expected 'mock' to be called once. Called 0 times.

../../src/ros2/rclpy/rclpy/test/test_subscription.py:194: AssertionError
```


The errors:
- [projectroot.test_client](https://build.ros2.org/job/Rci__nightly-connext_ubuntu_noble_amd64/531/testReport/junit/(root)/projectroot/test_client/)
- [projectroot.test_service](https://build.ros2.org/job/Rci__nightly-connext_ubuntu_noble_amd64/531/testReport/junit/(root)/projectroot/test_service/)

Might be related to this issue, although I'm not sure, because those are timeouts

i can reproduce this issue in my local environment.

```console
root@tomoyafujita-B760M-Pro-RS-D4:~/ros2_ws/colcon_ws# colcon test --event-handlers console_direct+ --packages-select rclpy --ctest-args -R test_subscription
Starting >>> rclpy   
...
37: ../../src/ros2/rclpy/rclpy/test/test_subscription.py ................F.. [ 82%]
37: ....                                                                     [100%]
37: 
37: =================================== FAILURES ===================================
37: _________________________ test_on_new_message_callback _________________________
37: 
37: self = <Mock id='126426529552032'>, args = (1,), kwargs = {}
37: msg = "Expected 'mock' to be called once. Called 0 times."
37: 
37:     def assert_called_once_with(self, /, *args, **kwargs):
37:         """assert that the mock was called exactly once and that that call was
37:         with the specified arguments."""
37:         if not self.call_count == 1:
37:             msg = ("Expected '%s' to be called once. Called %s times.%s"
37:                    % (self._mock_name or 'mock',
37:                       self.call_count,
37:                       self._calls_repr()))
37: >           raise AssertionError(msg)
37: E           AssertionError: Expected 'mock' to be called once. Called 0 times.
37: 
37: /usr/lib/python3.12/unittest/mock.py:955: AssertionError
37: 
37: During handling of the above exception, another exception occurred:
37: 
37: test_node = <rclpy.node.Node object at 0x72fbf8624500>
37: 
37:     def test_on_new_message_callback(test_node) -> None:
37:         topic_name = '/topic'
37:         cb = Mock()
37:         sub = test_node.create_subscription(
37:             msg_type=Empty,
37:             topic=topic_name,
37:             qos_profile=10,
37:             callback=cb)
37:         pub = test_node.create_publisher(Empty, topic_name, 10)
37:         sub.handle.set_on_new_message_callback(cb)
37:         cb.assert_not_called()
37:         pub.publish(Empty())
37: >       cb.assert_called_once_with(1)
37: E       AssertionError: Expected 'mock' to be called once. Called 0 times.
37: 
37: ../../src/ros2/rclpy/rclpy/test/test_subscription.py:196: AssertionError
37: ---------------------------- Captured stdout setup -----------------------------
37: RTI Connext DDS Non-commercial license is for academic, research, evaluation and personal use only. USE FOR COMMERCIAL PURPOSES IS PROHIBITED. See RTI_LICENSE.TXT for terms. Download free tools at rti.com/ncl. License issued to Non-Commercial User license@rti.com For non-production use only.
37: Expires on 00-jan-00
37: Please contact support@rti.com with any questions or comments.
37: - generated xml file: /root/ros2_ws/colcon_ws/build/rclpy/test_results/rclpy/test_subscription.xunit.xml -
37: =========================== short test summary info ============================
37: FAILED ../../src/ros2/rclpy/rclpy/test/test_subscription.py::test_on_new_message_callback
37: ========================= 1 failed, 22 passed in 4.83s =========================
37: -- run_test.py: return code 1
37: -- run_test.py: verify result file '/root/ros2_ws/colcon_ws/build/rclpy/test_results/rclpy/test_subscription.xunit.xml'
1/1 Test #37: test_subscription ................***Failed    5.25 sec
```

--- a/rclpy/test/test_subscription.py
+++ b/rclpy/test/test_subscription.py
@@ -190,13 +190,31 @@ def test_on_new_message_callback(test_node) -> None:
         qos_profile=10,
         callback=cb)
     pub = test_node.create_publisher(Empty, topic_name, 10)
+
+    # Wait for publisher and subscriber to discover each other
+    max_seconds_to_wait = 5
+    end_time = time.time() + max_seconds_to_wait
+    while sub.get_publisher_count() != 1:
+        time.sleep(0.1)
+        assert time.time() <= end_time  # timeout waiting for pub/sub to discover each other
+
     sub.handle.set_on_new_message_callback(cb)
     cb.assert_not_called()
     pub.publish(Empty())
+
+    # Wait for the callback to be invoked
+    end_time = time.time() + max_seconds_to_wait
+    while cb.call_count == 0 and time.time() <= end_time:
+        time.sleep(0.1)
+
     cb.assert_called_once_with(1)
     sub.handle.clear_on_new_message_callback()
+    cb.reset_mock()
     pub.publish(Empty())
-    cb.assert_called_once()
+
+    # Wait a bit to ensure the message would have been received if callback was still set
+    time.sleep(1.0)
+    cb.assert_not_called()
 
 
 def test_subscription_set_content_filter(test_node) -> None:

